os/fs/smartfs/smartfs_utils.c

```diff
@@ -1716,139 +1962,255 @@ int smartfs_countdirentries(struct smartfs_mountpt_s *fs, struct smartfs_entry_s
+
+	/* 'data_len' represents length of DATA in a sector, the chainheader is not counted in file data */
+	data_len = fs->fs_llformat.availbytes - sizeof(struct smartfs_chain_header_s);
+	/* Number of bytes to be retained in the last sector after shrinking file */
+	last_data = (length % data_len);
+
+	savefilepos = sf->filepos;
+	/* Seek till point 'length' of the file, file pointer lies at position of requested 'length' now */
+	smartfs_seek_internal(fs, sf, length, SEEK_SET);
+	if (length < savefilepos) {
+		savefilepos = length;
+	}
 
-	/* Walk through the directory's sectors and count entries */
+	offset = sizeof(struct smartfs_chain_header_s) + last_data;
+	/* Keep as many sectors as needed and replace extra bytes in the last needed sector with ERASEDSTATE */
+#ifdef CONFIG_SMARTFS_USE_SECTOR_BUFFER
+	memset(&sf->buffer[offset], CONFIG_SMARTFS_ERASEDSTATE, fs->fs_llformat.availbytes - offset);
+	chainheader = (struct smartfs_chain_header_s *)sf->buffer;
+	*(uint16_t *)chainheader->used = last_data;
+	*(uint16_t *)chainheader->nextsector = SMARTFS_ERASEDSTATE_16BIT;
+
+	/* New file length = requested length. Filepos is already at position 'length' */
+	sf->entry.datalen = length;
+	smartfs_sync_internal(fs, sf);
+#else
+	smartfs_setbuffer(&readwrite, sf->currsector, 0, fs->fs_llformat.availbytes, (FAR uint8_t *)fs->fs_rwbuffer);
+	ret = FS_IOCTL(fs, BIOC_READSECT, (unsigned long)&readwrite);
+	if (ret < 0) {
+		fdbg("Failed to read file sector no. %d, error = %d\n", nextsector, ret);
+		goto errout_with_filepos;
+	}
 
-	nextsector = entry->firstsector;
-	header = (struct smartfs_chain_header_s *)fs->fs_rwbuffer;
+	chainheader = (struct smartfs_chain_header_s *)fs->fs_rwbuffer;
+	memset(&fs->fs_rwbuffer[offset], CONFIG_SMARTFS_ERASEDSTATE, fs->fs_llformat.availbytes - offset);
+	*(uint16_t *)chainheader->used = last_data;
+	*(uint16_t *)chainheader->nextsector = SMARTFS_ERASEDSTATE_16BIT;
 
-	while (nextsector != SMARTFS_ERASEDSTATE_16BIT) {
-		/* Read the next sector's header into our buffer */
+	sf->entry.datalen = length;
+	ret = FS_IOCTL(fs, BIOC_WRITESECT, (unsigned long)&readwrite);
+	if (ret < 0) {
+		fdbg("Unable to write to sector %d while shrinking file, error = %d\n", readwrite.logsector, ret);
+		goto errout_with_filepos;
+	}
+#endif
 
-		smartfs_setbuffer(&readwrite, nextsector, 0, sizeof(struct smartfs_chain_header_s), (uint8_t *)fs->fs_rwbuffer);
+	nextsector = SMARTFS_NEXTSECTOR(chainheader);
+	chainheader = (struct smartfs_chain_header_s *)fs->fs_rwbuffer;
+
+	while (nextsector != SMARTFS_ERASEDSTATE_16BIT) {
+		smartfs_setbuffer(&readwrite, nextsector, 0, sizeof(struct smartfs_chain_header_s), (FAR uint8_t *)fs->fs_rwbuffer);
 		ret = FS_IOCTL(fs, BIOC_READSECT, (unsigned long)&readwrite);
 		if (ret < 0) {
-			fdbg("Error reading sector %d header, ret : %d\n", readwrite.logsector, ret);
-			return ret;
+			fdbg("Unable to read sector %d, error = %d\n", readwrite.logsector, ret);
+			goto errout_with_filepos;
```

Comment:
I think 'break' will be enough..

Reply:
Both will work the same.
But dont you think using goto errout is better since we are taking care of an error returned case?

Reply:
goto should be used only when absolutely necessary.
Because
```
1. There's chance that it can make spaghetti code.
2. Some of compiler takes more time when it optimize code.
```
But it has advantage, you can use it as a c version of try-catch, please check other code, it should be used to handle exceptional error.

Reply:
Done

Reply:
Not changed.

Reply:
Hmm yes,
New function code, will change.